[YUNIKORN-204] Basic controller logic

kingamarton opened the pull request:
This is a WIP PR. There are a couple of missing thins to to: adding tests, fix the TODO's, but I am creating it because some feedback it would be very helpful regarding the main idea. (especially because there are some things what smells a bit hacky for me, but I had no better idea).
@wilfred-s, @yangwwei, @HuangTing-Yao can you please check it when you will have some time?

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Stop() {
+	// noop
```
this is not a noop. it needs to stop the controller

kingamarton replied:
Can you please elaborate on what exactly do you expect about stopping the controller? Closing the channel, or something else?

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+	"go.uber.org/zap"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
```
do we really need to depend on this package?

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+}
+
+type Controller struct {
+	kubeClient       kubernetes.Interface
+	extensionsClient apiextensionsclient.Interface
+	appClient        applicationclient.Interface
+	informer         cache.SharedIndexInformer
+	lister           applister.ApplicationLister
```
is it possible to move most of these things to `apifactory.go`? 
the `APIProvider` is supposed to provide a convenient way to access client libraries.

kingamarton replied:
I will remove kubeclient and extensionsClient from here, since I need them only for initialising the other variables. The other ones are all application related and I would keep them here, since we need them to be created only if the CRD feature is turned on.

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+	})
+	informerFactory.Start(wait.NeverStop)
+
+	utilruntime.Must(appv1.AddToScheme(appscheme.Scheme))
```
what is this for?

kingamarton replied:
this utilruntime.Must call is handling the errors. Actually it will panic if the scheme registration is failing. I've seen this pattern used in other places, when starting the controller, so I followed this one as well.

yangwwei replied:
Got it, thx

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+}
+
+func (c *Controller) Run() {
+	//TODO: handle the case when the crd is not registered: register it programatically
+	defer utilruntime.HandleCrash()
```
I am not quite sure what does this mean.
why do we need to handle the case the CRD is not registered? If the CRD is not registered, I think we should not even start the scheduler. And also, what does the `HandleCrash()` do here?

kingamarton replied:
Right now we have two places where we can enable the CRD: the scheduler conf and the helm chart. In my opinion we should reduce it to only one place, the scheduler conf. If we will kepp this two places, I think it will be a frequent user error that they will enable it in one place and forget to enable it in the helm chart for example, or will forget to register it manually. I think in this case it will be useful to register it from the code.

yangwwei replied:
OK, if this is the common pattern, then that is fine.

kingamarton replied:
It is not a common pattern, it is just easy to forget to enable it in both places. I created a different Jira for it: YUNIKORN-346

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+}
+
+func (appMgr *AppManager) getAppMetadata(app *appv1.Application) (interfaces.ApplicationMetadata, bool) {
+	appID := app.Name
```
Like I commented here: https://github.com/apache/incubator-yunikorn-k8shim/pull/152#discussion_r453930011.
> According to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/. Object UID should be unique across the cluster. Maybe we can use that as the applicationID. That means the user only needs to specify a namespace-unique name for each app-CRD. Also, metadata has generateName field, so if the user doesn't specify any name, K8s will generate one with that as the prefix. 

I think we should let `appID` = spec.UID, not the name, to ensure it is unique.
We can put the appName in the tag, in the format of `namespace/appName`

yangwwei replied:
Could you please fix this based on the comment here: https://issues.apache.org/jira/browse/YUNIKORN-340?focusedCommentId=17169061&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17169061

kingamarton replied:
Since using UID is more complicated, I will use namespace-name as applicationID

yangwwei replied:
Make sense to me, thanks

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+	}, true
+}
+
+func convertShimAppStateToAppCRDState(status string) appv1.ApplicationStateType {
+	switch status {
+	case corecache.New.String():
+		return appv1.NewApplicationState
+	case corecache.Accepted.String():
+		return appv1.AcceptedState
+	case corecache.Starting.String():
+		return appv1.StartingState
+	case corecache.Running.String():
+		return appv1.RunningState
+	case corecache.Waiting.String():
+		return appv1.WaitingState
+	case corecache.Rejected.String():
+		return appv1.RejectedState
+	case corecache.Completed.String():
+		return appv1.CompletedState
+	case corecache.Killed.String():
+		return appv1.KilledState
+	default:
+		return "Undefined"
+	}
```
I think this is one of the main things we need to fix.
We need to make shim side app state in sync with the core side states, pls take a look at the core side status: http://yunikorn.apache.org/docs/next/design/scheduler_object_states#application-state.
I would even suggest creating a separate JIRA to get this fixed first.
Ideally, all the state transition are done in the core side, and core leverages UpdateApplication API to send the state updates to the shim, and that triggers the transition of shim side app-state-machine.

kingamarton replied:
I completely agree with you that we should keep the states in synch. For me this one seems a bigger refactoring task. And also we have different statuses for the shim side and different ones for the core side:
https://github.com/apache/incubator-yunikorn-k8shim/blob/master/pkg/cache/application.go#L71-L101
https://github.com/apache/incubator-yunikorn-core/blob/master/pkg/cache/application_state.go#L70-L100
If we want refactor it now, after then it will be easier to handle the AppCRD status as well, than it is now.

yangwwei replied:
Based on the offline discussion, we will not do the refactoring for now given the complexity of this problem.

yangwwei commented on pkg/appmgmt/sparkoperator/spark.go:
```diff
@@ -104,6 +105,12 @@ func (os *Manager) Stop() {
 	os.stopCh <- struct{}{}
 }
 
+func (os *Manager) HandleCallbackEvents() func(obj interface{}) {
+	return func(obj interface{}) {
+		//noop
+	}
+}
+
```
this can be removed once the handler function is moved to `AppManagementService`.

yangwwei commented on pkg/cache/application_events.go:
```diff
@@ -47,6 +47,35 @@ func (st SimpleApplicationEvent) GetApplicationID() string {
+type ApplicationStatusChangeEvent struct {
+	applicationID string
+	event         events.ApplicationEventType
+	State         string
```
for a state change, it needs to have 2 fields: `source` and `destination` state.
if we do not have this in the scheduler-interface, I think we need to add that too. Sorry for missing this earlier.

kingamarton replied:
From the application point of view we don't need the source state, because when this event is generated the transition is already done, so we just need to update it in the CRD. The destination field will be unused. I suggest to extend it with the destination only if we will need it.

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	corecache "github.com/apache/incubator-yunikorn-core/pkg/cache"
```
please do not depend on this package

kingamarton replied:
removed

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+
+func convertShimAppStateToAppCRDState(status string) appv1.ApplicationStateType {
+	switch status {
+	case corecache.New.String():
```
we should not depend on the core side package directly,
based on current limitations, I think we will just do the parse with strings...

kingamarton replied:
removed

yangwwei commented on pkg/shim/scheduler.go:
```diff
@@ -110,10 +111,21 @@ func newShimSchedulerInternal(ctx *cache.Context, apiFactory client.APIProvider,
 	dispatcher.RegisterEventHandler(dispatcher.EventTypeTask, ctx.TaskEventHandler())
 	dispatcher.RegisterEventHandler(dispatcher.EventTypeNode, ctx.SchedulerNodeEventHandler())
 	dispatcher.RegisterEventHandler(dispatcher.EventTypeScheduler, ss.SchedulerEventHandler())
+	ss.registerAppManagerHandler(constants.AppManagerHandlerName)
```
Like the previous comment, please replace this with 

```
dispatcher.RegisterEventHandler(dispatcher. EventTypeAppStateUpdate, am.ApplicationStateUpdateEventHandler())
```

yangwwei commented on pkg/shim/scheduler.go:
```diff
@@ -110,10 +111,21 @@ func newShimSchedulerInternal(ctx *cache.Context, apiFactory client.APIProvider,
 	return ss
 }
 
+func (ss *KubernetesShim) registerAppManagerHandler (appMgrName string) {
+	ss.context.AddSchedulingEventHandlers()
+	appMgr := ss.appManager.GetManagerByName(appMgrName)
+	if appMgr != nil {
+		dispatcher.RegisterEventHandler(dispatcher.EventTypeAppStatus, appMgr.HandleCallbackEvents())
+	} else {
+		log.Logger.Debug("App manager not registered",
+		zap.String("app manager name", appMgrName))
+	}
+}
```
this can be removed

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+	apiextensionsClient := apiextensionsclient.NewForConfigOrDie(config)
+	appClient := applicationclient.NewForConfigOrDie(config)
+
+	informerFactory := appinformers.NewSharedInformerFactory(appClient, time.Minute*1)
+	informer := informerFactory.Apache().V1alpha1().Applications()
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    appMgr.addApp,
+		UpdateFunc: appMgr.updateApp,
+		DeleteFunc: appMgr.deleteApp,
+	})
+	informerFactory.Start(wait.NeverStop)
```
you have a `Run()` function, we should move the start of the informerFactory stuff to there.

kingamarton replied:
I am thinking about removing this line, because it seems to be enough to start the informers

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+		timeoutCh <- struct{}{}
+	}()
+	if ok := cache.WaitForCacheSync(timeoutCh, c.informer.HasSynced); !ok {
+		log.Logger.Error("Timeout expired during waiting for caches to sync. Skip starting the application controller")
```
this log is wrong, it is not skipping the controller start, everything has been started. this only means the controller is running without the up-to-date sync'd cache.

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+Update the application in the scheduler
+*/
+func (appMgr *AppManager) updateApp(oldObj interface{}, newObj interface{}) {
+	newApp, ok := newObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("newObj is not an Application")
+	}
+	oldApp, ok := oldObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("oldObj is not an Application")
+	}
+	// No need to update if ResourceVersion is not changed
+	if newApp.ResourceVersion == oldApp.ResourceVersion {
+		log.Logger.Info("No need to update because app object is not modified")
+		return
+	}
+
+	// handle spec change
+	if !equality.Semantic.DeepEqual(oldApp.Spec, newApp.Spec) {
+		//TODO: handle update, discuss what should be updated since only the queue is the common property yet
+	}
```
I'd prefer to leave it empty, with a comment

```
// not implement today
```

kingamarton replied:
If we will not implement the update, I would rather delete it than leaving it empty

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+func (appMgr *AppManager) deleteApp(obj interface{}) {
+	app, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
```
forgot a return after the error logging here?

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+	appID := app.Name
+	err := appMgr.amProtocol.RemoveApplication(appID)
+	if err != nil {
+		log.Logger.Error("Application removal failed",
+			zap.String("appID", appID),
+			zap.Error(err))
```
I think a `return` needs to be added after the error log

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+func (appMgr *AppManager) addApp(obj interface{}) {
+	appCRD, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
```
I think a return needs to be added after the error log

yangwwei commented on pkg/controller/application/app_controller.go:
```diff
@@ -0,0 +1,295 @@
+		if app == nil {
+			appMgr.amProtocol.AddApplication(&interfaces.AddApplicationRequest{
+				Metadata: appMeta,
+				//TODO: debug recovery
+				//TODO: try to define the owner reference for the submitted pods to this CRD and try of we delete the
+				//CRD the pods will be deleted as well
```
We need a separate JIRA to track how to implement the recovery, pls file one

kingamarton replied:
Filed YUNIKORN-347

yangwwei commented on pkg/dispatcher/dispatcher.go:
```diff
@@ -179,7 +180,11 @@ func Start() {
 				switch v := event.(type) {
 				case events.ApplicationEvent:
-					getEventHandler(EventTypeApp)(v)
+					if v.GetEvent() == events.AppStateChange {
```
I feel this could be misleading.
Can we define another interface for the app state change? In `pkg/common/events/events.go`, create another interface specifically for handling the app-state-updates from the core to the shim.

kingamarton replied:
Created a new interface: events.ApplicationStatusEvent

kingamarton commented on pkg/appmgmt/interfaces/appmgr.go:
```diff
@@ -42,4 +42,6 @@ type AppManager interface {
 	// the stop() function is called.
 	Stop()
+
+	HandleCallbackEvents() func(obj interface{})
```
done

kingamarton commented:
@yangwwei I have addressed your review comments. I still have to add the unit tests and try to set the ownerReferrence for the pods belonging to an application.
@wilfred-s  can you please take a look to this changes as well?

yangwwei commented on pkg/client/clients.go:
```diff
@@ -67,7 +71,8 @@ func (c *Clients) WaitForSync(interval time.Duration, timeout time.Duration) err
 			c.ConfigMapInformer.Informer().HasSynced() &&
-			c.NamespaceInformer.Informer().HasSynced()
+			c.NamespaceInformer.Informer().HasSynced() &&
+			c.ApplicationInformer.Informer().HasSynced()
```
Seems this could be causing a NPE when app-CRD is not enabled.
Could you please add a nil check here?

kingamarton replied:
Good catch. This is valid. I will fix it

yangwwei commented:
hi @kingamarton changes look good to me. But could you please check the travis job failure? Thanks.

kingamarton commented:
@yangwwei, the actual failure is because I need to change the go.mod file to point to the latest interface and also to the latest core. For updating it we need to merge the core side changes first otherwise the shim and the core will point to different interface versions.

yangwwei commented:
hi @kingamarton  can you at least fix the `codecov/project` error? Thanks

yangwwei commented:
hi, @kingamarton thanks for getting this done. Now it looks good.
The latest job shows only the codecov/project is failing, but I don't think that is caused by your patch. We will need to improve the code coverage in the shim repo to get those places, such as `application_events.go`. We can do that in follow up tasks, I have created https://issues.apache.org/jira/browse/YUNIKORN-383 for this.
